### ISSUE TITLE
TakePictureAsync: image "always" stored in landscape even when taken in portrait

### ISSUE THREAD

**Issue by ahmedalejo** [jamesmontemagno/Xamarin.Plugins#134]
Using [version 1.0.2](https://www.nuget.org/packages/Xam.Plugin.Media/1.0.2)

There´s also a Xamarin.Forms Camera sample from here - https://github.com/XForms/Xamarin-Forms-Labs-Samples/tree/master/XF.Labs.CameraSample I am able to select or take a photo.

```csharp
private async Task SelectPicture()
{
    mediaPicker = DependencyService.Get<IMediaPicker>();
    imageSource = null;
        var mediaFile = await mediaPicker.SelectPhotoAsync(new CameraMediaStorageOptions
            {
                DefaultCamera = CameraDevice.Front,
                MaxPixelDimension = 400
            });
        imageSource = ImageSource.FromStream(() => mediaFile.Source);
        img.Source  = imageSource;

}

private async Task TakePictureAsync()
{
    mediaPicker = DependencyService.Get<IMediaPicker>();
    imageSource = null;
        var mediaFile = await mediaPicker.TakePhotoAsync(new CameraMediaStorageOptions
            {
                DefaultCamera = CameraDevice.Front,
                MaxPixelDimension = 400
            });
        imageSource = ImageSource.FromStream(() => mediaFile.Source);
        img.Source  = imageSource;

}
```
the code for the image is simply
```csharp
    img = new Image
    {
        BackgroundColor = Color.White,
        Aspect = Aspect.AspectFit
    };    
```
There are a couple of issues:

First one. You can take a photo or select a stored one and it will then show it on the page. If you select a photo it displays it correctly, either portrait or landscape. When you take a photo, it only displays in landscape mode, so if the image was taken in portrait, the image shows on the side. This isn't catastrophic, but it would be better to show the image how it was taken.

I have only tried this on Android so far. Does anyone have any ideas on how I can solve the above problems?

REFERENCE:
[**StackOverflows** Xamarin.Forms Take photo with camera shows wrong orientation and crashes on back button](http://stackoverflow.com/questions/26296889/xamarin-forms-take-photo-with-camera-shows-wrong-orientation-and-crashes-on-back)

FIX:
[**Fixing the rotation**](http://www.androidzeitgeist.com/2013/01/fixing-rotation-camera-picture.html)

**Comment by jamesmontemagno** [comment]
I think that this is just based on the device and not always the case.

**Comment by thekiwi** [comment]
Yes, that's what I've experienced too - it works as expected on my Nexus 4, but portrait photos are incorrectly rotated on my Samsung S6.

**Comment by ahmedalejo** [comment]
@jamesmontemagno, like @thekiwi said this is mostly an issue on Samsung devices:

which currently has the highest market shares by TestCloud:
![image](https://cloud.githubusercontent.com/assets/7012074/11625543/dcc5cf4c-9cc2-11e5-8938-d7f93977959d.png)

**Comment by ahmedalejo** [comment]
```csharp
using System;
using System.Threading.Tasks;
using Android.Media;
using MediaOrientation = Android.Media.Orientation;
using System.IO;
using Android.Graphics;

namespace Mobile.Platform.Media
{
///Shared
    public interface IMediaFileExtensions
    {
        /// <summary>
        ///  Does nothing
        /// </summary>
        /// <param name="file">The file image</param>
        /// <returns>True if rotation occured, else false</returns>
        Task<bool> FixOrientationAsync(IMediaFile file);
    }

#if __ANDROID__
[assembly: Dependency(typeof(MediaFileExtensions))]
#else
[assembly: Dependency(typeof(MediaFileExtension.NoopMediaFileExtensions))]
#endif

///Android specific
    public class MediaFileExtensions : IMediaFileExtensions
    {
        /// <summary>
        ///  Rotate an image if required.
        /// </summary>
        /// <param name="file">The file image</param>
        /// <returns>True if rotation occured, else fal</returns>
        async public Task<bool> FixOrientationAsync(IMediaFile file)
        {
            if (file == null)
                return false;
            try
            {

                var filePath = file.Path;
                var orientation = GetRotation(filePath);

                if (!orientation.HasValue)
                    return false;

                Bitmap bmp = RotateImage(filePath, orientation.Value);
                var quality = 90;

                using (var stream = File.Open(filePath, FileMode.OpenOrCreate))
                    await bmp.CompressAsync(Bitmap.CompressFormat.Png, quality, stream);

                bmp.Recycle();

                return true;
            }
            catch (Exception ex)
            {                               
                ex.Report(); //Extension method for Xamarin.Insights
                return false;
            }
        }

        static int? GetRotation(string filePath)
        {
            try
            {
                ExifInterface ei = new ExifInterface(filePath);
                var orientation = (MediaOrientation)ei.GetAttributeInt(ExifInterface.TagOrientation, (int)MediaOrientation.Normal);
                switch (orientation)
                {
                    case MediaOrientation.Rotate90:
                        return 90;
                    case MediaOrientation.Rotate180:
                        return 180;
                    case MediaOrientation.Rotate270:
                        return 270;
                    default:
                        return null;
                }

            }
            catch (Exception ex)
            {
                ex.Report();
                return null;
            }
        }

        private static Bitmap RotateImage(string filePath, int rotation)
        {
            Bitmap originalImage = BitmapFactory.DecodeFile(filePath);

            Matrix matrix = new Matrix();
            matrix.PostRotate(rotation);
            var rotatedImage = Bitmap.CreateBitmap(originalImage, 0, 0, originalImage.Width, originalImage.Height, matrix, true);
            originalImage.Recycle();
            return rotatedImage;
        }
    }
}

```

**Comment by highplains** [comment]
+1 this would be handy!

**Comment by praveenmohanmm** [comment]
hi,
i am also facing the portrait issue. is there work around for that.
i am using samsung galaxy s6.

thanks in advance.

**Comment by ahmedalejo** [comment]
@praveenmohanmm the workaround was stated above,

**Comment by nunohorta** [comment]
@ahmedalejo any workaround for iOS? this is happening to me on the iPad

**Comment by highplains** [comment]
Also a lot of images that are stored in the gallery on the phones dont have Exif info which makes that workaround not complete

**Comment by ahmedalejo** [comment]
@nunohorta i am not familiar with iOS issue. Could you please file an issue detailing your issue.

**Comment by ahmedalejo** [comment]
@highplains this fix/workaround is for pictures taken by the device´s camera. i.e **TakePhotoAsync** and not **SelectPhotoAsync** which refers to picking:
> a lot of images that are stored in the gallery

but if you encounter an issue with this workaround then i´d be glad to resolve it. 

Cheers!!

**Comment by jamesmontemagno** [comment]
Will be in 2.4.0-beta1

**Comment by sairamb87** [comment]
2.4.0 - beta1, when will this version be available in NuGet ?

**Comment by jamesmontemagno** [comment]
Nugets are already available latest is here: https://www.nuget.org/packages/Xam.Plugin.Media/2.5.1-beta40

**Comment by alequeshow** [comment]
@ahmedalejo. I've tried to implement the workaround but I can't get how you refer the android implementation in PCL project. I just can't resolve the MediaFileExtensions in the piece below:

`#if __ANDROID__
[assembly: Dependency(typeof(MediaFileExtensions))]`

It happens mainly because it's a cyclic dependency error and if I place the android specific in another project, I just can't refer it in PCL as well. I wonder how should I resolve this dependency issue. Any suggestions?

By the way, I got the alpha version and nothing has changed in my portrait pictures, so that's why I'm trying to use the workaround.

And thanks for the solution!
